### PR TITLE
fix(sleep): use uucore's from_str to support parsing hex

### LINKED ISSUE #7669: sleep: Duration parsing does not support all float formats (e.g. hex)
Same as #7475, but for `sleep`. `sleep` isn't using the common duration parser, instead, it's using a parser from another crate (`fundu`)

GNU coreutils supports things like hex formatting:

```
sleep 0x1.1
sleep 0x1.1p-1
sleep 0x10.1
```

Not a huge deal, probably a pretty rare edge case that I doubt anybody cares about, but still worth sharing code with `timeout`.

### PR DESCRIPTION
**Pull request by Qelxiros**
closes #7669

`sleep` now uses uucore's `from_str` instead of `fundu` to support hex parsing. This changes some error output (although it looks to be more in line with GNU, not less), so I changed the relevant tests as well. However, at first glance, it looks like it might be returning "empty string" in unexpected cases (e.g. '\n').

### CODE REVIEW

**Comment by drinkcat** [comment]
@Qelxiros Please rebase, #7648 made changes in this code recently.

File: src/uucore/src/lib/parser/parse_time.rs [hunk 1]
```diff
@@ -49,16 +51,21 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
     if len == 0 {
         return Err("empty string".to_owned());
```
Comment (drinkcat) [inline src/uucore/src/lib/parser/parse_time.rs:52]: I think we should return `invalid time interval {}` here too, for consistency with coreutils.

File: src/uucore/src/lib/parser/parse_time.rs [hunk 2]
```diff
@@ -49,16 +51,21 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
...
-        's' => (slice, 1),
-        'm' => (slice, 60),
-        'h' => (slice, 60 * 60),
-        'd' => (slice, 60 * 60 * 24),
-        val if !val.is_alphabetic() => (string, 1),
+    let (numstr, times) = match (string.chars().next_back().unwrap(), base) {
+        ('s', _) => (slice, 1),
+        ('m', _) => (slice, 60),
+        ('h', _) => (slice, 60 * 60),
+        // if the number is hex, GNU coreutils will parse the trailing d as part of the number, not as days
```
Comment (drinkcat) [inline src/uucore/src/lib/parser/parse_time.rs:66]: Yeah, this is #7671. This might be tricky to handle properly, you might want to do that as a separate PR.

Comment (drinkcat) [reply src/uucore/src/lib/parser/parse_time.rs:66]: I feel the cleanest way to handle this would be to directly call into `num_parser::parse`, e.g. by adding a list of allowed suffixes (the multiplication can be done outside of `parse` though).

Also, we'd want to disallow stuff like `sleep \'1` (which is currently parsed as ascii `1` == 49 seconds)

File: src/uucore/src/lib/features/parser/parse_time.rs [hunk 3]
```diff
@@ -57,20 +57,27 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
...
+    };
+    let (numstr, times) = match (trimmed.chars().next_back().unwrap(), base) {
+        ('s', _) => (slice, 1),
+        ('m', _) => (slice, 60),
+        ('h', _) => (slice, 60 * 60),
+        ('d', 10) => (slice, 60 * 60 * 24),
+        ('d', 16) if slice.contains('p') => (slice, 60 * 60 * 24),
+        ('d', 16) if !slice.contains('p') => (trimmed, 1),
+        (val, _) if !val.is_alphabetic() => (trimmed, 1),
+        _ => match trimmed.to_ascii_lowercase().as_str() {
```
Comment (drinkcat) [inline src/uucore/src/lib/features/parser/parse_time.rs:80]: BTW, please add tests in this file for all of those corner cases (hex digits with suffix, hex float with suffix, etc. etc.)

File: src/uucore/src/lib/features/parser/num_parser.rs [hunk 4]
```diff
@@ -356,7 +356,7 @@ fn construct_extended_big_decimal<'a>(
 // TODO: As highlighted by clippy, this function _is_ high cognitive complexity, jumps
 // around between integer and float parsing, and should be split in multiple parts.
 #[allow(clippy::cognitive_complexity)]
-fn parse(
+pub(crate) fn parse(
```
Comment (drinkcat) [inline src/uucore/src/lib/features/parser/num_parser.rs:359]: Not needed right now?

File: src/uucore/src/lib/features/parser/parse_time.rs [hunk 5]
```diff
@@ -57,20 +57,27 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
     // https://github.com/rust-lang/rust/issues/57391
     const NANOSECOND_DURATION: Duration = Duration::from_nanos(1);
 
-    let len = string.len();
+    let trimmed = string.trim();
```
[the remark this answers is not in the document]
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs:60]: I did this because the original sleep code passed a `trim`med version to fundu. Thanks for the insight here!

File: src/uucore/src/lib/features/parser/parse_time.rs [hunk 6]
```diff
@@ -57,20 +57,27 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
...
     };
-    let (numstr, times) = match string.chars().next_back().unwrap() {
-        's' => (slice, 1),
-        'm' => (slice, 60),
-        'h' => (slice, 60 * 60),
-        'd' => (slice, 60 * 60 * 24),
-        val if !val.is_alphabetic() => (string, 1),
-        _ => match string.to_ascii_lowercase().as_str() {
+    let base = match trimmed.get(0..2) {
+        Some("0x") => 16,
```
[the remark this answers is not in the document]
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs:69]: Good call on the edge cases. I saw that spot in `parse`, but I let my hubris get the better of me.

File: src/uucore/src/lib/features/parser/parse_time.rs (hunk 3 above)
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs:80]: Will do!

File: src/uucore/src/lib/features/parser/parse_time.rs [hunk 7]
```diff
@@ -57,20 +57,27 @@ pub fn from_str(string: &str) -> Result<Duration, String> {
...
+        _ => 10,
+    };
+    let (numstr, times) = match (trimmed.chars().next_back().unwrap(), base) {
+        ('s', _) => (slice, 1),
+        ('m', _) => (slice, 60),
+        ('h', _) => (slice, 60 * 60),
+        ('d', 10) => (slice, 60 * 60 * 24),
+        ('d', 16) if slice.contains('p') => (slice, 60 * 60 * 24),
+        ('d', 16) if !slice.contains('p') => (trimmed, 1),
```
[the remark this answers is not in the document]
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs:78]: It seems as though I'll have to do some form of base checking in `from_str` so that I know which suffixes to pass to `parse`. If I do something like `string[1..=2].contains(['x', 'X'])` (accounting for out-of-bounds errors), does that address all the edge cases? Or did I misunderstand what you meant by allowed suffixes?

[the remark this answers is not in the document]
Comment (drinkcat) [reply src/uucore/src/lib/features/parser/parse_time.rs:78]: I'd modify the `parse` function to handle the duration suffixes (depending on a parameter, just like the function can either parse float or integer depending on the parameter), it'd be so much easier to do it there (instead of trying to work around the issue by doing some pre-parsing of the number to figure out if the suffix `d` is meant to be a day or part of an hex number.

Since `parse` is greedy, it'll happily parse all of the digits of an hex number (or any number really), _then_ when you finish parsing, you can check if the leftover character is just a `s`/`m`/`h`/`d` and multiple the value there (or return a PartialMatch if there is some other trailing characters).

[the remark this answers is not in the document]
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs:78]: The problem I see with that is that I'm stuck with one of these two cases: either `from_str` saw the trailing `d` and erroneously multiplies by `24 * 60 * 60`, or I modify `parse` to return which suffix it found (if any), which seems like a significant change, particularly to a function that's so widely used. A scan over the string for an `x|X` doesn't seem too expensive, and it avoids contributing more to the monstrosity of cognitive complexity that is the `parse` function. All that said, if you prefer this way, I'll do it.

File: src/uucore/src/lib/parser/parse_time.rs (hunk 2 above)
Comment (Qelxiros) [reply src/uucore/src/lib/parser/parse_time.rs]: Just out of curiosity, what's the use case for `\'1`?

File: src/uucore/src/lib/features/parser/parse_time.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (drinkcat) [reply src/uucore/src/lib/features/parser/parse_time.rs:78]: First, agree about the monstrosity of `parse` ,-)

But... from looking into parsing logic over the last few weeks, I'm really worried about corner cases (the more I look, the more I find...). I could find a few broken examples off the top of my head here, but I cannot guarantee I covered everything, so I'd really prefer having the logic in one place, so that future changes apply to all the parsing in all the utilities.

And, I think the change in `parse` doesn't need to be very big...

- First, I'd change `parse` to take an enum as parameter, that can be Decimal, Integral, or Duration.
- Then, towards the end of `parse`, right before `let ebd_result ==`, you can check if there is a valid suffix, and pre-multiply `digits` (and now that I look at the code, you'd need to do something similar in `parse_special_value` since `infd` is a valid duration, but maybe it's not too bad...).
- Bonus: `// Parse the " and ' prefixes separately` can be ignored for Duration parameter! (yet another special case you'd need to handle in `parse_time` otherwise... (#7678)

File: src/uucore/src/lib/parser/parse_time.rs (hunk 2 above)
Comment (drinkcat) [reply src/uucore/src/lib/parser/parse_time.rs]: To print ASCII/unicode values:
```
printf "%d %d %x\n" \'1 \'a \'謝
49 97 8b1d
```

(`seq` also accepts that value, maybe it's harder to find a use case ,-P)

File: src/uucore/src/lib/features/parser/parse_time.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/parse_time.rs]: Thanks so much for your help and your patience! My latest commit should be good, but I'll make sure to fix any problems if they appear during CI.

File: src/uucore/src/lib/features/parser/num_parser.rs [hunk 8]
```diff
@@ -253,7 +254,14 @@ fn parse_special_value(
             if negative {
                 special = -special;
             }
-            let match_len = str.len();
+            let mut match_len = str.len();
```
Comment (drinkcat) [inline src/uucore/src/lib/features/parser/num_parser.rs:257]: Oh I was commenting on a previous version earlier?! This is good.

File: src/uucore/src/lib/features/parser/num_parser.rs [hunk 9]
```diff
@@ -353,24 +361,34 @@ fn construct_extended_big_decimal<'a>(
     Ok(ExtendedBigDecimal::BigDecimal(bd))
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
```
Comment (drinkcat) [inline src/uucore/src/lib/features/parser/num_parser.rs:364]: `PartialEq` doesn't feel necessary?

File: src/uucore/src/lib/features/parser/num_parser.rs [hunk 10]
```diff
@@ -230,14 +230,15 @@ impl ExtendedParser for ExtendedBigDecimal {
...
         input: &str,
     ) -> Result<ExtendedBigDecimal, ExtendedParserError<'_, ExtendedBigDecimal>> {
-        parse(input, false)
+        parse(input, ParseTarget::Decimal, &[])
     }
 }
 
-fn parse_special_value(
-    input: &str,
+fn parse_special_value<'a>(
```
Comment (drinkcat) [inline src/uucore/src/lib/features/parser/num_parser.rs:237]: I'm still a Rust noob, but why do you need those lifetimes? (I also wonder if it should be only on input and not allowed_suffixes)

File: tests/by-util/test_sleep.rs [hunk 11]
```diff
@@ -273,7 +281,32 @@ fn test_negative_interval() {
...
+#[case::frac("0x0.1")]
+#[case::frac_suffix("0x0.1s")]
+#[case::frac_suffix("0x0.001h")]
+#[case::scientific("0x1.0p-3")]
+#[case::scientific_suffix("0x1.0p-4s")]
+fn test_valid_hex_duration(#[case] input: &str) {
+    new_ucmd!().args(&["--", input]).succeeds().no_output();
+}
+
+#[rstest]
```
Comment (drinkcat) [inline tests/by-util/test_sleep.rs:301]: Please also add test cases for capitalized duration (e.g. `1S` is invalid, `InFD` is invalid, but `INFd` is valid)

**Review by drinkcat — COMMENTED** [review]
Thank you!! I think it looks quite good (I was afraid it'd look worse than that ,-P)! Just a few nits.

**Comment by drinkcat** [comment]
Oh, one more thing, please squash your commits into logical units (1 big commit would be ok, or maybe a couple: one for `num_parser` and one for `parse_time`), so that the git history stays somewhat clean.

File: src/uucore/src/lib/features/parser/num_parser.rs (hunk 9 above)
Comment (Qelxiros) [reply src/uucore/src/lib/features/parser/num_parser.rs:364]: `PartialEq` is used for comparison. There's an argument that I should remove the `Eq` derivation, but it doesn't actually change the implementation. I can't remove only `PartialEq` because it's required by `Eq`.

**Comment by Qelxiros** [comment]
@drinkcat There should be an option to "squash and merge", which will combine all of my changes into one commit (possibly two with a merge commit). If it's easier for you, I'm happy to make one commit on my end.

File: src/uucore/src/lib/features/parser/num_parser.rs (hunk 9 above)
Comment (drinkcat) [reply src/uucore/src/lib/features/parser/num_parser.rs:364]: Gotcha thanks.

File: src/uucore/src/lib/features/parser/num_parser.rs (hunk 10 above)
Comment (drinkcat) [reply src/uucore/src/lib/features/parser/num_parser.rs:237]: Sounds good thanks.

**Comment by drinkcat** [comment]
Thanks! I think this looks good now.

> @drinkcat There should be an option to "squash and merge", which will combine all of my changes into one commit (possibly two with a merge commit). If it's easier for you, I'm happy to make one commit on my end.

I'll let @sylvestre or @cakebaker comment.

**Comment by Qelxiros** [comment]
I've confirmed that this would close #7678, but the relevant tests aren't there yet. This implementation of `from_str` also has the potential to close #7670.

However, I don't want this PR to get too much scope creep. Should I add these two features here, or should I wait for this PR to get merged and then branch off of main for the other two?

**Comment by RenjiSann** [comment]
> There should be an option to "squash and merge", which will combine all of my changes into one commit (possibly two with a merge commit). If it's easier for you, I'm happy to make one commit on my end.

As a good practice, I usually make 2 commits (unless I'm pushing a massive refactor which needs separate commits) : One for the changes, one for the tests. This is not possible to do using Github's interface, hence the request to squash the commits on your side.

Thank you !

**Comment by sylvestre** [comment]
> Congrats! The gnu test tests/misc/sleep is no longer failing!

kudos :)

**Comment by RenjiSann** [comment]
Thanks !

**Comment by sylvestre** [comment]
@Qelxiros as you touched this, maybe you could also remove the fundu dependency here too: https://github.com/uutils/coreutils/blob/82f354fb16cf710674679d4df836847e41d4352b/src/uu/tail/src/args.rs#L231
